source/opt/flatten_array_matrix_stage_var.cpp

```diff
@@ -0,0 +1,908 @@
+
+  NestedCompositeComponents flattened_vars;
+  while (column_count > 0) {
+    NestedCompositeComponents flattened_vars_for_column =
+        CreateFlattenedStageVarsForReplacement(column_type, storage_class,
+                                               extra_array_length);
+    flattened_vars.AddComponent(flattened_vars_for_column);
+    --column_count;
+  }
+  return flattened_vars;
+}
+
+FlattenArrayMatrixStageVariable::NestedCompositeComponents
+FlattenArrayMatrixStageVariable::CreateFlattenedStageVarsForReplacement(
+    Instruction* stage_var_type, SpvStorageClass storage_class,
+    uint32_t extra_array_length) {
+  // Handle array case.
+  if (stage_var_type->opcode() == SpvOpTypeArray) {
+    return CreateFlattenedStageVarsForArray(stage_var_type, storage_class,
+                                            extra_array_length);
+  }
+
+  // Handle matrix case.
+  if (stage_var_type->opcode() == SpvOpTypeMatrix) {
+    return CreateFlattenedStageVarsForMatrix(stage_var_type, storage_class,
+                                             extra_array_length);
+  }
+
+  // Handle scalar or vector case.
+  NestedCompositeComponents flattened_var;
+  uint32_t type_id = stage_var_type->result_id();
+  if (extra_array_length != 0) {
+    type_id = GetArrayType(type_id, extra_array_length);
+  }
+  uint32_t ptr_type_id =
+      context()->get_type_mgr()->FindPointerToType(type_id, storage_class);
+  uint32_t id = TakeNextId();
+  std::unique_ptr<Instruction> variable(
+      new Instruction(context(), SpvOpVariable, ptr_type_id, id,
+                      std::initializer_list<Operand>{
+                          {SPV_OPERAND_TYPE_STORAGE_CLASS,
+                           {static_cast<uint32_t>(storage_class)}}}));
+  flattened_var.SetSingleComponentVariable(variable.get());
+  context()->AddGlobalValue(std::move(variable));
+  return flattened_var;
+}
+
+Instruction* FlattenArrayMatrixStageVariable::GetTypeOfStageVariable(
+    Instruction* stage_var, bool has_extra_arrayness) {
+  uint32_t pointee_type_id = GetPointeeTypeIdOfVar(stage_var);
+
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  Instruction* type_inst = def_use_mgr->GetDef(pointee_type_id);
+  if (has_extra_arrayness) {
+    assert(type_inst->opcode() == SpvOpTypeArray &&
+           "Stage variable with an extra arrayness must be an array");
+    // Get the type without extra arrayness.
+    uint32_t elem_type_id =
+        type_inst->GetSingleWordInOperand(kOpTypeArrayElemTypeInOperandIndex);
+    type_inst = def_use_mgr->GetDef(elem_type_id);
+  }
+
+  return type_inst;
+}
+
+Instruction* FlattenArrayMatrixStageVariable::GetStageVariable(
+    uint32_t stage_var_id) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  Instruction* stage_var = def_use_mgr->GetDef(stage_var_id);
+  if (stage_var == nullptr) {
+    std::string message("Stage variable does not exist");
```

Comment:
I do not see a test where this is actually output.  From what I understand flow on how this is called, I don't think this could ever be executed.  I suggest turning this into an assert. If I am wrong, please add a test where this code is executed.